File: projects/Grocery-Order-System-Java/GroceryOrder.java
/**
 * This class represents a grocery order. It can hold any grocery items,
 * create a standard or random order, and calculate totals and savings.
 * 
 * @author Sanskar Luitel
 * @version Project 10
 * Lab Section Tue/Thur 1:05-2:45PM
 */
import java.util.*;

public class GroceryOrder {

    private ArrayList<GroceryItem> items;

    public GroceryOrder(boolean standard) {
        items = new ArrayList<>();
        if (standard) {
            createStandardOrder();
        } else {
            createRandomOrder();
        }
    }

    private void createStandardOrder() {
        Apple a = new Apple();
        a.setCount(5);
        items.add(a);

        items.add(new Chips());

        Donut d1 = new Donut(true);
        d1.setCount(4);
        items.add(d1);

        Donut d2 = new Donut(false);
        d2.setCount(3);
        items.add(d2);

        items.add(new Eggs());
        items.add(new Eggs());

        Orange o = new Orange(true);
        o.setCount(3);
        items.add(o);

        items.add(new RotisserieChicken(true));
        items.add(new RotisserieChicken(false));
    }

    private void createRandomOrder() {
        Random rand = new Random();
        int numItems = rand.nextInt(17) + 4; 

        for (int i = 0; i < numItems; i++) {
            int choice = rand.nextInt(6);

            switch (choice) {
                case 0 -> { 
                    Apple a = new Apple();
                    a.setCount(rand.nextInt(5) + 1);
                    items.add(a);
                }
                case 1 -> items.add(new Chips());
                case 2 -> { 
                    boolean frosted = rand.nextBoolean();
                    Donut d = new Donut(frosted);
                    d.setCount(rand.nextInt(12) + 1);
                    items.add(d);
                }
                case 3 -> items.add(new Eggs());
                case 4 -> { 
                    boolean organic = rand.nextBoolean();
                    Orange o = new Orange(organic);
                    o.setCount(rand.nextInt(3) + 1);
                    items.add(o);
                }
                case 5 -> { 
                    boolean hot = rand.nextBoolean();
                    items.add(new RotisserieChicken(hot));
                }
            }
        }
    }

    public double getTotalFullPrice() {
        double total = 0;
        for (GroceryItem g : items) {
            total += g.getFullPrice();
        }
        return total;
    }

    public double getTotalDiscountPrice() {
        double total = 0;
        for (GroceryItem g : items) {
            if (g instanceof DiscountedItem di) {
                total += di.getDiscountPrice();
            } else {
                total += g.getFullPrice();
            }
        }
        return total;
    }

    public double getSavings() {
        return getTotalFullPrice() - getTotalDiscountPrice();
    }

    public ArrayList<GroceryItem> getAllItems() {
        return items;
    }

    public ArrayList<CountableItem> getCountableItems() {
        ArrayList<CountableItem> list = new ArrayList<>();
        for (GroceryItem g : items) {
            if (g instanceof CountableItem c) {
                list.add(c);
            }
        }
        return list;
    }

    public ArrayList<DiscountedItem> getDiscountedItems() {
        ArrayList<DiscountedItem> list = new ArrayList<>();
        for (GroceryItem g : items) {
            if (g instanceof DiscountedItem d) {
                list.add(d);
            }
        }
        return list;
    }
}

File: projects/Grocery-Order-System-CSharp/proj11/GroceryOrder.cs
using System;
using System.Collections.Generic;

public class GroceryOrder
{
    private List<IGroceryItem> _items = new List<IGroceryItem>();

    public GroceryOrder(bool standard)
    {
        if (standard)
            CreateStandardOrder();
        else
            CreateRandomOrder();
    }

    private void CreateStandardOrder()
    {
        _items.Add(new Apple() { Count = 5 });
        _items.Add(new Chips());
        _items.Add(new Donut(true) { Count = 4 });
        _items.Add(new Donut(false) { Count = 3 });
        _items.Add(new Eggs());
        _items.Add(new Eggs());
        _items.Add(new Orange(true) { Count = 3 });
        _items.Add(new RotisserieChicken(true));
        _items.Add(new RotisserieChicken(false));
    }

    private void CreateRandomOrder()
    {
        Random rand = new Random();
        int numItems = rand.Next(4, 21);

        for (int i = 0; i < numItems; i++)
        {
            int choice = rand.Next(0, 6);
            if (choice == 0)
                _items.Add(new Apple() { Count = rand.Next(1, 6) });
            else if (choice == 1)
                _items.Add(new Chips());
            else if (choice == 2)
                _items.Add(new Donut(rand.Next(0, 2) == 1) { Count = rand.Next(1, 13) });
            else if (choice == 3)
                _items.Add(new Eggs());
            else if (choice == 4)
                _items.Add(new Orange(rand.Next(0, 2) == 1) { Count = rand.Next(1, 4) });
            else
                _items.Add(new RotisserieChicken(rand.Next(0, 2) == 1));
        }
    }

    public double TotalFullPrice
    {
        get
        {
            double total = 0;
            foreach (IGroceryItem g in _items)
                total += g.FullPrice;
            return total;
        }
    }

    public double TotalDiscountPrice
    {
        get
        {
            double total = 0;
            foreach (IGroceryItem g in _items)
            {
                IDiscountedItem d = g as IDiscountedItem;
                total += d != null ? d.DiscountPrice : g.FullPrice;
            }
            return total;
        }
    }

    public double Savings => TotalFullPrice - TotalDiscountPrice;

    public List<IGroceryItem> AllItems => _items;

    public List<ICountableItem> CountableItems
    {
        get
        {
            List<ICountableItem> list = new List<ICountableItem>();
            foreach (IGroceryItem g in _items)
            {
                ICountableItem c = g as ICountableItem;
                if (c != null) list.Add(c);
            }
            return list;
        }
    }

    public List<IDiscountedItem> DiscountedItems
    {
        get
        {
            List<IDiscountedItem> list = new List<IDiscountedItem>();
            foreach (IGroceryItem g in _items)
            {
                IDiscountedItem d = g as IDiscountedItem;
                if (d != null) list.Add(d);
            }
            return list;
        }
    }
}

